### DISCUSSION TITLE
Support for more complex geometries?

### DISCUSSION THREAD

**Discussion by Benj-Ward** [zwicker-group/py-pde#418] [General]
Hello Dr. Zwicker.

I am studying some applications of the diffusion equation, and for the systems that I am interested in modeling, I will need some more complicated geometries. My first question, is it possible to create a hole in the mesh of the simulation? Or would it be possible to import a mesh from another software such as gmesh? 

This is definitely a stretch, but is there support for the immersed boundary method, or a way to approximate it? Less of a stretch and as effective, are you familiar with a way to define boundary conditions on arbitrary planes in the solution domain or approximate them in some way? 

I suspect that the answer to much of the above will be no, but something that I know will work from a technical perspective is defining a paired PDE for the diffusion coefficient, which will solve for D = 0 at certain places and D = 1 at others (and rapidly switch between) to get dc/dt = 0 where D = 0. What do you think about this from a mathematical point of view as an implicit approximation for a desired condition of del(c) = 0? Thanks.

**Comment by hpmeyer25** [comment]
I think I have a similar application need. I'm interested in modeling diffusion in cylindrical geometries with multiple layers in the r axis. I've been able to do this for monolithic cyliniders using pde.DiffusionPDE from the documentation. However, the Cylindrical Symmetry Grid does not seem to support periodic boundary onditions on the r-axis. Am I missing something in the documentation for implementing this?

Alternatively if applying the approximation proposed by the OP, can you recommend a suitable way to pass D as a function of the r or z coordinates to pde.DiffusionPDE?

**Reply by david-zwicker** [reply]
The `DiffusionPDE` class itself does not support spatially dependent diffusivities, but the more general `PDE` class does! We discussed how to do this in https://github.com/zwicker-group/py-pde/discussions/229 and [there is also an example](https://py-pde.readthedocs.io/en/latest/examples_gallery/index.html).

I don't understand what you mean by "the Cylindrical Symmetry Grid does not seem to support periodic boundary onditions on the r-axis" – how would periodic boundary conditions work on a cylindrical grid?

**Reply by hpmeyer25** [reply]
Hi Dr. Zwicker,

Thank you very much for your reply! I will try with the `PDE` class based on the example as you suggested.

The situation I'd like to model is a multilayer assembly (a cylinder surrounded by am annulus, both infinite in z). I think the interface  between the inner cylinder and the annulus (r=a)  needs an equal flux condition (dC(a-)/dr = dC(a+)/dr), as well as a condition relating the concentration at the interface of the layers (e.g., C(a-) = C(a+) for simple cases or C(a+) = f(C(a-) for more complex cases). Is there a way to implement this with the CylindricalSymmetryGrid class?
![image](https://github.com/zwicker-group/py-pde/assets/34461390/d57c464e-2770-45be-a73d-137664e83536)

I've tried using the `from_bounds` method, but `from_bound(bounds, shape, periodic = [True, False])` does not seem to enable this, as the documentation indicates that the first item in the `periodic` argument is ignored.

**Reply by david-zwicker** [reply]
Cylindrical annuli (so a cylinder with a finite inner radius) are not currently supported, although it is generally possible to add this support (`PolarSymGrid` and `SphericalSymGrid` support it, for instance). However, this would require quite some work on the `CylindricalSymGrid` class.

**Reply by david-zwicker** [reply]
It turned out to be easier to implement than I expected, so I just merged #420 into the `master` branch, which supports symmetric cylindrical grids with an inner radius.

**Comment by david-zwicker** [comment]
Unfortunately, complicated geometries are outside the scope of `py-pde`, which focuses on simple geometries that can be described by finite differences. If you need to investigate complex geometries, you should instead try one of the many finite element or finite volume methods.
